### ISSUE TITLE
How can I custom loss function during fine-tuning?[Question] 

### ISSUE THREAD

**Issue by qingyunyanran** [haotian-liu/LLaVA#1636]
### Question

I can't find where the model defined loss function. And I want to custom loss function but I am afraid that if I just add this function in the tuning script it might destroy the whole model.

**Comment by anas-zafar** [comment]
@qingyunyanran were you able to figure this out? Thanks

**Comment by AnnaGao0827** [comment]
Hi, I have the same question. When finetuing the LLaVA model, can we use custom loss fuction and data format? Or we must use Language Modeling loss & human-gpt data format? Thanks!

**Comment by formidify** [comment]
To customize loss, you can just override the compute_loss function in the LLaVATrainer class in llava/train/llava_trainer.py. The original compute_loss function can be found here: https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/trainer.py#L3353

If you need to customize the input data format, you can do so in train.py. Hope this helps!

**Comment by AnnaGao0827** [comment]
> To customize loss, you can just override the compute_loss function in the LLaVATrainer class in llava/train/llava_trainer.py. The original compute_loss function can be found here: https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/trainer.py#L3353
> 
> If you need to customize the input data format, you can do so in train.py. Hope this helps!

Thank you sooo much. Your guidance helps a lot!

**Comment by 1835969208** [comment]
> 要自定义损失，您只需覆盖 llava/train/llava_trainer.py 中 LLaVATrainer 类中的 compute_loss 函数。原始 compute_loss 函数可在此处找到：https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/trainer.py#L3353
> 
> 如果需要自定义输入数据格式，可以在 train.py 中执行此作。希望这有帮助！

@formidify 但我没法在LLaVATrainer 类中找到 compute_loss 函数，LLaVATrainer 类中内容如下：
class LLaVATrainer(Trainer):

    def _get_train_sampler(self) -> Optional[torch.utils.data.Sampler]:
        if self.train_dataset is None or not has_length(self.train_dataset):
            return None

        if self.args.group_by_modality_length:
            lengths = self.train_dataset.modality_lengths
            return LengthGroupedSampler(
                self.args.train_batch_size,
                world_size=self.args.world_size * self.args.gradient_accumulation_steps,
                lengths=lengths,
                group_by_modality=True,
            )
        else:
            return super()._get_train_sampler()

    def create_optimizer(self):
        """
        Setup the optimizer.

        We provide a reasonable default that works well. If you want to use something else, you can pass a tuple in the
        Trainer's init through `optimizers`, or subclass and override this method in a subclass.
        """
        if is_sagemaker_mp_enabled():
            return super().create_optimizer()

        opt_model = self.model

        if self.optimizer is None:
            decay_parameters = get_parameter_names(opt_model, ALL_LAYERNORM_LAYERS)
            decay_parameters = [name for name in decay_parameters if "bias" not in name]
            if self.args.mm_projector_lr is not None:
                projector_parameters = [name for name, _ in opt_model.named_parameters() if "mm_projector" in name]
                optimizer_grouped_parameters = [
                    {
                        "params": [
                            p for n, p in opt_model.named_parameters() if (n in decay_parameters and n not in projector_parameters and p.requires_grad)
                        ],
                        "weight_decay": self.args.weight_decay,
                    },
                    {
                        "params": [
                            p for n, p in opt_model.named_parameters() if (n not in decay_parameters and n not in projector_parameters and p.requires_grad)
                        ],
                        "weight_decay": 0.0,
                    },
                    {
                        "params": [
                            p for n, p in opt_model.named_parameters() if (n in decay_parameters and n in projector_parameters and p.requires_grad)
                        ],
                        "weight_decay": self.args.weight_decay,
                        "lr": self.args.mm_projector_lr,
                    },
                    {
                        "params": [
                            p for n, p in opt_model.named_parameters() if (n not in decay_parameters and n in projector_parameters and p.requires_grad)
                        ],
                        "weight_decay": 0.0,
                        "lr": self.args.mm_projector_lr,
                    },
                ]
            else:
                optimizer_grouped_parameters = [
                    {
                        "params": [
                            p for n, p in opt_model.named_parameters() if (n in decay_parameters and p.requires_grad)
                        ],
                        "weight_decay": self.args.weight_decay,
                    },
                    {
                        "params": [
                            p for n, p in opt_model.named_parameters() if (n not in decay_parameters and p.requires_grad)
                        ],
                        "weight_decay": 0.0,
                    },
                ]

            optimizer_cls, optimizer_kwargs = Trainer.get_optimizer_cls_and_kwargs(self.args)

            self.optimizer = optimizer_cls(optimizer_grouped_parameters, **optimizer_kwargs)
            if optimizer_cls.__name__ == "Adam8bit":
                import bitsandbytes

                manager = bitsandbytes.optim.GlobalOptimManager.get_instance()

                skipped = 0
                for module in opt_model.modules():
                    if isinstance(module, nn.Embedding):
                        skipped += sum({p.data_ptr(): p.numel() for p in module.parameters()}.values())
                        logger.info(f"skipped {module}: {skipped/2**20}M params")
                        manager.register_module_override(module, "weight", {"optim_bits": 32})
                        logger.debug(f"bitsandbytes: will optimize {module} in fp32")
                logger.info(f"skipped: {skipped/2**20}M params")

        return self.optimizer

    def _save_checkpoint(self, model, trial, metrics=None):
        if getattr(self.args, 'tune_mm_mlp_adapter', False):
            from transformers.trainer_utils import PREFIX_CHECKPOINT_DIR
            checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"

            run_dir = self._get_output_dir(trial=trial)
            output_dir = os.path.join(run_dir, checkpoint_folder)

            # Only save Adapter
            keys_to_match = ['mm_projector', 'vision_resampler']
            if getattr(self.args, "use_im_start_end", False):
                keys_to_match.extend(['embed_tokens', 'embed_in'])

            weight_to_save = get_mm_adapter_state_maybe_zero_3(self.model.named_parameters(), keys_to_match)

            if self.args.local_rank == 0 or self.args.local_rank == -1:
                self.model.config.save_pretrained(output_dir)
                torch.save(weight_to_save, os.path.join(output_dir, f'mm_projector.bin'))
        else:
            super(LLaVATrainer, self)._save_checkpoint(model, trial, metrics)

    def _save(self, output_dir: Optional[str] = None, state_dict=None):
        if getattr(self.args, 'tune_mm_mlp_adapter', False):
            pass
        else:
            super(LLaVATrainer, self)._save(output_dir, state_dict)

**Comment by formidify** [comment]
You are not going to find it in llava_trainer.py. LLaVATrainer is an instance of the Transformers library class Trainer, so you need to override it based on the original class definition from https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/trainer.py#L3353

**Comment by 1835969208** [comment]
非常感谢你的回答和帮助，我想具体的问一下，您的意思是 我需要在llava/train/llava_trainer.py 中 LLaVATrainer 中自定义一个和Transformers 库类 Trainer中一样的compute_loss函数，还是我要将https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/trainer.py#L3353 中完整的trainer复制过来再修改里面的compute_loss函数。再次感谢你的帮助，期待你的回答

**Comment by formidify** [comment]
You just need to define a compute_loss function in your LLaVATrainer class to override the original.

**Comment by 1835969208** [comment]
@formidify 非常感谢你的回答，十分明确清晰，祝你工作顺利
